### ISSUE TITLE
Building a Hand-Drawn Digit Recognizer with PyTorch and MNIST

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11754]
Building a Hand-Drawn Digit Recognizer with PyTorch and MNIST<br>
<br>
https://ift.tt/iKdYCj3<br>
<br>
Zachary Proser<br>
<br>

<div><div><div><div><div>
<img alt="The MNIST mind mapper" src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fmnist-mind-mapper-splash.c59d9dcf.webp&amp;w=3840&amp;q=75"><h2>Table of contents</h2>
<h2>Demo of the "MNIST mind-mapper" app in action</h2>
<p>You can watch this short demo to see how the app works:</p>

<h2>Overview</h2>
<p>I've been an application and infrastructure developer for the majority of my career, so I wanted to get hands-on with training and deploying a neural network.</p>
<img alt="A neural network" src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fpytorch-mnist-neural-network.111bc3e0.webp&amp;w=3840&amp;q=75"><p>I also wanted to wrap that trained neural network in a REST API, so that I could build a frontend that would allow folks
to play with it, because interacting with something is more engaging than reading a text description of it.</p>
<p>I knew it would be important to go beyond the working neural net, because issues often arise at the seams, when you're fitting system components together.</p>
<p>This article details the steps I took and the many issues I encountered along the way to building and successfully deploying my original vision.</p>
<h2>Open source code</h2>
<p>I open source most of my work, and this project is no exception:</p>
<h2>App flow diagram</h2>
<p>Let's step through how the app works, end to end:</p>
<img alt="The MNIST mind mapper" src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fmnist-mind-mapper-flow.518b13fb.webp&amp;w=1200&amp;q=75"><p>The frontend exposes a small drawable canvas, which the user scribbles on.</p>
<p>On a regular interval, the frontend captures what the user drew, using the <code>toDataURL</code> method:</p>
<pre><code> 
 <span>toDataURL</span><span>(</span>type<span>?</span><span>:</span> <span>string</span><span>,</span> quality<span>?</span><span>:</span> <span>any</span><span>)</span><span>:</span> <span>string</span><span>;</span>
</code></pre>
<p>This image is sent to the backend API, which wraps the trained neural network. The backend runs inference on the image, and returns the predicted digit, which the frontend displays.</p>
<h2>Step 1: Model training</h2>
<p>The following code is all that's needed with pytorch to:</p>
<ol>
<li>Define a simple neural network architecture</li>
<li>Load the MNIST dataset</li>
<li>Train the neural network</li>
<li>Save the model to disk so that it can be loaded and reused later</li>
</ol>
<pre><code><span>import</span> torch
<span>import</span> torch<span>.</span>nn <span>as</span> nn
<span>import</span> torch<span>.</span>optim <span>as</span> optim
<span>from</span> torchvision <span>import</span> datasets<span>,</span> transforms
<span>from</span> torch<span>.</span>utils<span>.</span>data <span>import</span> DataLoader

<span>class</span> <span>SimpleNN</span><span>(</span>nn<span>.</span>Module<span>)</span><span>:</span>
    <span>def</span> <span>__init__</span><span>(</span>self<span>)</span><span>:</span>
        <span>super</span><span>(</span>SimpleNN<span>,</span> self<span>)</span><span>.</span>__init__<span>(</span><span>)</span>
        self<span>.</span>flatten <span>=</span> nn<span>.</span>Flatten<span>(</span><span>)</span>
        self<span>.</span>fc1 <span>=</span> nn<span>.</span>Linear<span>(</span><span>28</span> <span>*</span> <span>28</span><span>,</span> <span>128</span><span>)</span>
        self<span>.</span>relu <span>=</span> nn<span>.</span>ReLU<span>(</span><span>)</span>
        self<span>.</span>fc2 <span>=</span> nn<span>.</span>Linear<span>(</span><span>128</span><span>,</span> <span>10</span><span>)</span>

    <span>def</span> <span>forward</span><span>(</span>self<span>,</span> x<span>)</span><span>:</span>
        x <span>=</span> self<span>.</span>flatten<span>(</span>x<span>)</span>
        x <span>=</span> self<span>.</span>fc1<span>(</span>x<span>)</span>
        x <span>=</span> self<span>.</span>relu<span>(</span>x<span>)</span>
        x <span>=</span> self<span>.</span>fc2<span>(</span>x<span>)</span>
        <span>return</span> x

transform <span>=</span> transforms<span>.</span>Compose<span>(</span><span>[</span>transforms<span>.</span>ToTensor<span>(</span><span>)</span><span>,</span> transforms<span>.</span>Normalize<span>(</span><span>(</span><span>0.1307</span><span>,</span><span>)</span><span>,</span> <span>(</span><span>0.3081</span><span>,</span><span>)</span><span>)</span><span>]</span><span>)</span>

train_dataset <span>=</span> datasets<span>.</span>MNIST<span>(</span>root<span>=</span><span>'./data'</span><span>,</span> train<span>=</span><span>True</span><span>,</span> download<span>=</span><span>True</span><span>,</span> transform<span>=</span>transform<span>)</span>
test_dataset <span>=</span> datasets<span>.</span>MNIST<span>(</span>root<span>=</span><span>'./data'</span><span>,</span> train<span>=</span><span>False</span><span>,</span> download<span>=</span><span>True</span><span>,</span> transform<span>=</span>transform<span>)</span>

train_loader <span>=</span> DataLoader<span>(</span>train_dataset<span>,</span> batch_size<span>=</span><span>64</span><span>,</span> shuffle<span>=</span><span>True</span><span>)</span>
test_loader <span>=</span> DataLoader<span>(</span>test_dataset<span>,</span> batch_size<span>=</span><span>1000</span><span>,</span> shuffle<span>=</span><span>False</span><span>)</span>

device <span>=</span> torch<span>.</span>device<span>(</span><span>"cuda"</span> <span>if</span> torch<span>.</span>cuda<span>.</span>is_available<span>(</span><span>)</span> <span>else</span> <span>"cpu"</span><span>)</span>
model <span>=</span> SimpleNN<span>(</span><span>)</span><span>.</span>to<span>(</span>device<span>)</span>
criterion <span>=</span> nn<span>.</span>CrossEntropyLoss<span>(</span><span>)</span>
optimizer <span>=</span> optim<span>.</span>Adam<span>(</span>model<span>.</span>parameters<span>(</span><span>)</span><span>,</span> lr<span>=</span><span>0.001</span><span>)</span>

num_epochs <span>=</span> <span>5</span>

<span>for</span> epoch <span>in</span> <span>range</span><span>(</span>num_epochs<span>)</span><span>:</span>
    model<span>.</span>train<span>(</span><span>)</span>
    <span>for</span> batch_idx<span>,</span> <span>(</span>data<span>,</span> target<span>)</span> <span>in</span> <span>enumerate</span><span>(</span>train_loader<span>)</span><span>:</span>
        data<span>,</span> target <span>=</span> data<span>.</span>to<span>(</span>device<span>)</span><span>,</span> target<span>.</span>to<span>(</span>device<span>)</span>
        optimizer<span>.</span>zero_grad<span>(</span><span>)</span>
        output <span>=</span> model<span>(</span>data<span>)</span>
        loss <span>=</span> criterion<span>(</span>output<span>,</span> target<span>)</span>
        loss<span>.</span>backward<span>(</span><span>)</span>
        optimizer<span>.</span>step<span>(</span><span>)</span>
        
        <span>if</span> batch_idx <span>%</span> <span>100</span> <span>==</span> <span>0</span><span>:</span>
            <span>print</span><span>(</span><span><span>f"Epoch </span><span><span>{</span>epoch<span>+</span><span>1</span><span>}</span></span><span>/</span><span><span>{</span>num_epochs<span>}</span></span><span>, Batch </span><span><span>{</span>batch_idx<span>}</span></span><span>/</span><span><span>{</span><span>len</span><span>(</span>train_loader<span>)</span><span>}</span></span><span>, Loss: </span><span><span>{</span>loss<span>.</span>item<span>(</span><span>)</span><span>:</span><span>.4f</span><span>}</span></span><span>"</span></span><span>)</span>

model<span>.</span><span>eval</span><span>(</span><span>)</span>
correct <span>=</span> <span>0</span>
total <span>=</span> <span>0</span>

<span>with</span> torch<span>.</span>no_grad<span>(</span><span>)</span><span>:</span>
    <span>for</span> data<span>,</span> target <span>in</span> test_loader<span>:</span>
        data<span>,</span> target <span>=</span> data<span>.</span>to<span>(</span>device<span>)</span><span>,</span> target<span>.</span>to<span>(</span>device<span>)</span>
        output <span>=</span> model<span>(</span>data<span>)</span>
        _<span>,</span> predicted <span>=</span> torch<span>.</span><span>max</span><span>(</span>output<span>.</span>data<span>,</span> <span>1</span><span>)</span>
        total <span>+=</span> target<span>.</span>size<span>(</span><span>0</span><span>)</span>
        correct <span>+=</span> <span>(</span>predicted <span>==</span> target<span>)</span><span>.</span><span>sum</span><span>(</span><span>)</span><span>.</span>item<span>(</span><span>)</span>

<span>print</span><span>(</span><span><span>f"Accuracy on test set: </span><span><span>{</span><span>100</span> <span>*</span> correct <span>/</span> total<span>:</span><span>.2f</span><span>}</span></span><span>%"</span></span><span>)</span>

torch<span>.</span>save<span>(</span>model<span>.</span>state_dict<span>(</span><span>)</span><span>,</span> <span>"mnist_model.pth"</span><span>)</span>

</code></pre>
<p>With the training complete, I wanted to quickly sanity check the trained network's performance on the intended task, so I created a simple FAST API server that exposes a <code>/test_images</code> route:</p>
<p>You can view the entire <code>main.py</code> file in <a href="https://github.com/zackproser/pytorch-mnist/blob/main/main.py">the repository</a>, but we'll examine the route itself here, which runs a self-test and shares
the output as a PNG image like so:</p>
<img alt="MNIST test image" src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fmnist-test-images.56ba61af.webp&amp;w=3840&amp;q=75"><pre><code>app <span>=</span> FastAPI<span>(</span><span>)</span>

<span>@app<span>.</span>get</span><span>(</span><span>"/test_images"</span><span>)</span>
<span>async</span> <span>def</span> <span>test_images</span><span>(</span><span>)</span><span>:</span>
    
    <span>from</span> torchvision <span>import</span> datasets
    test_dataset <span>=</span> datasets<span>.</span>MNIST<span>(</span>root<span>=</span><span>'./data'</span><span>,</span> train<span>=</span><span>False</span><span>,</span> download<span>=</span><span>True</span><span>,</span> transform<span>=</span>transform<span>)</span>
    test_loader <span>=</span> torch<span>.</span>utils<span>.</span>data<span>.</span>DataLoader<span>(</span>test_dataset<span>,</span> batch_size<span>=</span><span>10</span><span>,</span> shuffle<span>=</span><span>True</span><span>)</span>

    
    images<span>,</span> labels <span>=</span> <span>next</span><span>(</span><span>iter</span><span>(</span>test_loader<span>)</span><span>)</span>

    fig<span>,</span> axs <span>=</span> plt<span>.</span>subplots<span>(</span><span>2</span><span>,</span> <span>5</span><span>,</span> figsize<span>=</span><span>(</span><span>12</span><span>,</span> <span>6</span><span>)</span><span>)</span>
    fig<span>.</span>suptitle<span>(</span><span>'Sample Test Images with Predictions'</span><span>)</span>

    <span>for</span> i<span>,</span> <span>(</span>image<span>,</span> label<span>)</span> <span>in</span> <span>enumerate</span><span>(</span><span>zip</span><span>(</span>images<span>,</span> labels<span>)</span><span>)</span><span>:</span>
        output <span>=</span> model<span>(</span>image<span>.</span>unsqueeze<span>(</span><span>0</span><span>)</span><span>.</span>to<span>(</span>device<span>)</span><span>)</span>
        _<span>,</span> predicted <span>=</span> torch<span>.</span><span>max</span><span>(</span>output<span>.</span>data<span>,</span> <span>1</span><span>)</span>
        
        ax <span>=</span> axs<span>[</span>i <span>//</span> <span>5</span><span>,</span> i <span>%</span> <span>5</span><span>]</span>
        ax<span>.</span>imshow<span>(</span>image<span>.</span>squeeze<span>(</span><span>)</span><span>.</span>numpy<span>(</span><span>)</span><span>,</span> cmap<span>=</span><span>'gray'</span><span>)</span>
        ax<span>.</span>set_title<span>(</span><span><span>f'Pred: </span><span><span>{</span>predicted<span>.</span>item<span>(</span><span>)</span><span>}</span></span><span>, True: </span><span><span>{</span>label<span>.</span>item<span>(</span><span>)</span><span>}</span></span><span>'</span></span><span>)</span>
        ax<span>.</span>axis<span>(</span><span>'off'</span><span>)</span>

    plt<span>.</span>tight_layout<span>(</span><span>)</span>
    
    
    <span>with</span> tempfile<span>.</span>NamedTemporaryFile<span>(</span>delete<span>=</span><span>False</span><span>,</span> suffix<span>=</span><span>'.png'</span><span>)</span> <span>as</span> tmp<span>:</span>
        plt<span>.</span>savefig<span>(</span>tmp<span>.</span>name<span>,</span> <span>format</span><span>=</span><span>'png'</span><span>)</span>
        tmp_path <span>=</span> tmp<span>.</span>name

    plt<span>.</span>close<span>(</span><span>)</span>  
    
    <span>return</span> FileResponse<span>(</span>tmp_path<span>,</span> media_type<span>=</span><span>"image/png"</span><span>,</span> filename<span>=</span><span>"test_images.png"</span><span>)</span>
</code></pre>
<p>Now I had a trained and working neural network and I was ready to deploy it. Here's where all the fun began.</p>
<h2>First challenge: Vercel's Python support is immature</h2>
<p>Vercel's beta support for the Python runtime for use in backend functions is very exciting. I believe the ability to deploy a mixed app with a Next.js frontend and a Python backend has huge potential.</p>
<p>The Python ecosystem is rich with popular machine learning libraries, utilities and datasets. Meanwhile, JavaScript provides an excellent way to provide charts, graphs, data visualizations, games and other interactive representations of complex data models.</p>
<p>I'm personally very excited to build a ton of example applications with Python backends and TypeScript frontends on Vercel. But we're not quite there yet.</p>
<p>Unfortunately, the docs Vercel has for using their Python runtime are very sparse, the examples are light and most of the new concepts are not sufficiently explained. You have to read through the Python Vercel templates to understand how everything fits together.</p>
<p>Errors are also opaque. The showstopper for getting my Python backend deployed successfully on Vercel was an unintuitive error message: <code>data too long</code>.</p>
<img alt="Vercel's Python support is immature" src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fvercel-python-data-too-long.c9d42ff5.webp&amp;w=3840&amp;q=75"><p>I was pretty sure that pytorch and torchvision were likely blowing out the 4.5MB size limit on serverless functions, but there wasn't a great way to confirm this. My model/weights file was just under 400KB.</p>
<p>Things that you'd expect to be extensively documented, such as how to get around common bundling issues, or how to deploy a Python backend with PyTorch and a custom model, are also sorely needed.</p>
<p>You find extremely important details like this hiding out in 4 year old GitHub issue comment threads:</p>
<img src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fvercel-python-github-exclude-files.da647623.webp&amp;w=1920&amp;q=75"><h2>Converting to ONNX and trying again</h2>
<p><a href="https://onnx.ai" target="_blank">ONNX, (which stands for Open Neural Network Exchange)</a> is a fascinating project that defines a common language for machine learning models, to allow Cloud providers and developers to more easily
write and deploy models.</p>
<p>You can convert a PyTorch model to ONNX using the <code>torch.onnx.export</code> function. ONNX CAN sometimes reduce the size of the exported model, so I decided to give it a shot.</p>
<p>I also was happy to take any excuse to play around with ONNX after reading about it. After converting the model to the ONNX format, I tested everything again locally to ensure the app still
functioned as intended. The potential size savings from ONNX did not end up making a difference in my case, and I got the same error.</p>
<h2>Finding a Python backend host in Modal.com</h2>
<p>In researching platforms that host your Python code as cloud functions, I found modal.com which was lightweight and fast to set up.</p>
<img alt="Modal.com dashboard" src="https://zackproser.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fmodal-dashboard.7c86433a.webp&amp;w=3840&amp;q=75"><p>I ended up converting the local version of my Fast API Python backend to a file I named <code>modal_function.py</code> which I
then deployed like so:</p>
<p><code>modal deploy modal_function.py</code></p>
<p>Of course, my prefence would be for the backend to be hosted on Vercel, so that I could colocate the frontend and backend code and have a simpler deployment model, but modal.com ended up
being exactly what I wanted in the absence of that.</p>
<p>Modal.com lets you sign up quickly, install their CLI and deploy your Python code as a serverless function or a deployed REST API, or an ephemeral testing endpoint. So far, it's been great.</p>
<p>There is the issue of cold starts, because Modal will spin down your container when your service is not receiving traffic, but overall I've been impressed with the developer experience.</p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Zachary Proser <br>
<br>
August 22, 2024 at 03:57PM
